orchestrator/graphql/utils/modify_class.py

```diff
@@ -0,0 +1,142 @@
+from enum import IntEnum, StrEnum
+from functools import partial, reduce
+from operator import or_
+from typing import Any, get_args, get_origin
+
+import strawberry
+from more_itertools import consume, first_true, side_effect
+from pydantic import create_model
+from pydantic.fields import FieldInfo
+from strawberry.experimental.pydantic.conversion_types import StrawberryTypeFromPydantic
+
+from nwastdlib import const
+from orchestrator.domain.base import DomainModel
+from orchestrator.types import is_list_type, is_union_type
+from pydantic_forms.types import strEnum
+
+MatchFunc = Callable
+ActionFunc = Callable
+ModifyMap = Iterable[tuple[MatchFunc, ActionFunc]]
+
+
+def is_subclass_of(annotation: type, class_or_tuple: type | tuple[type]) -> bool:
+    return inspect.isclass(annotation) and issubclass(annotation, class_or_tuple)
+
+
+is_domain_model = partial(is_subclass_of, class_or_tuple=DomainModel)
+is_int_enum = partial(is_subclass_of, class_or_tuple=IntEnum)
+is_str_enum = partial(is_subclass_of, class_or_tuple=(strEnum, StrEnum))
+
+
+def is_optional_enum_of_type(annotation: type | None, is_type_enum: Callable) -> bool:
+    origin = get_origin(annotation)
+    if origin is not None and origin is types.UnionType:
+        args = get_args(annotation)
+        return len(args) == 2 and is_type_enum(args[0]) and args[1] is type(None)
+    return False
+
+
+is_optional_int_enum = partial(is_optional_enum_of_type, is_type_enum=is_int_enum)
+is_optional_str_enum = partial(is_optional_enum_of_type, is_type_enum=is_str_enum)
+
+
+__MAPPED_TYPES = {}
+
+
+def map_type(annotation: type | None, modify_map: ModifyMap) -> type | None:
+    if result := __MAPPED_TYPES.get(annotation):
+        return result
+
+    match first_true(modify_map, pred=lambda match: match[0](annotation)):
+        case _, action:
+            result = action(annotation, modify_map)
+        case _:
+            result = None
+
+    __MAPPED_TYPES[annotation] = result
+    return result
+
+
+def map_union_members(annotation: type | None, modify_map: ModifyMap) -> Iterator[type]:
+    for orig_member in get_args(annotation):
+        if mapped_member := map_type(orig_member, modify_map):
+            yield mapped_member
+        else:
+            yield orig_member
+
+
+def class_walker(klass: type[DomainModel], modify_map: ModifyMap) -> type[DomainModel]:
+
+    def map_field(fi: FieldInfo) -> None:
+        annotation = fi.annotation
+        if mapped_type := map_type(annotation, modify_map):
+            fi.annotation = mapped_type
+        if is_list_type(annotation):
+            orig_type = get_args(annotation)[0]
+            if mapped_type := map_type(orig_type, modify_map):
+                fi.annotation = list[mapped_type]  # type: ignore
+        if is_union_type(annotation) and not is_union_type(annotation, klass):
+            fi.annotation = reduce(or_, map_union_members(annotation, modify_map))
```

Comment:
I added this to handle unions, but this unfortunately doesn't solve the problem which lies deeper in Strawberry

Reply:
https://git.ia.surfsara.nl/netdev/automation/projects/orchestrator/-/issues/2005